src/raw/array.rs

```diff
@@ -0,0 +1,242 @@
+    /// ObjectId.
+    pub fn get_object_id(&self, index: usize) -> ValueAccessResult<ObjectId> {
+        self.get_with(index, ElementType::ObjectId, RawBson::as_object_id)
+    }
+
+    /// Gets the boolean at the given index or returns an error if the value at that index isn't a
+    /// boolean.
+    pub fn get_bool(&self, index: usize) -> ValueAccessResult<bool> {
+        self.get_with(index, ElementType::Boolean, RawBson::as_bool)
+    }
+
+    /// Gets the DateTime at the given index or returns an error if the value at that index isn't a
+    /// DateTime.
+    pub fn get_datetime(&self, index: usize) -> ValueAccessResult<DateTime> {
+        self.get_with(index, ElementType::DateTime, RawBson::as_datetime)
+    }
+
+    /// Gets a reference to the BSON regex at the given index or returns an error if the
+    /// value at that index isn't a regex.
+    pub fn get_regex(&self, index: usize) -> ValueAccessResult<RawRegex<'_>> {
+        self.get_with(index, ElementType::RegularExpression, RawBson::as_regex)
+    }
+
+    /// Gets a reference to the BSON timestamp at the given index or returns an error if the
+    /// value at that index isn't a timestamp.
+    pub fn get_timestamp(&self, index: usize) -> ValueAccessResult<Timestamp> {
+        self.get_with(index, ElementType::Timestamp, RawBson::as_timestamp)
+    }
+
+    /// Gets the BSON int32 at the given index or returns an error if the value at that index isn't
+    /// a 32-bit integer.
+    pub fn get_i32(&self, index: usize) -> ValueAccessResult<i32> {
+        self.get_with(index, ElementType::Int32, RawBson::as_i32)
+    }
+
+    /// Gets BSON int64 at the given index or returns an error if the value at that index isn't a
+    /// 64-bit integer.
+    pub fn get_i64(&self, index: usize) -> ValueAccessResult<i64> {
+        self.get_with(index, ElementType::Int64, RawBson::as_i64)
+    }
+
+    /// Gets a reference to the raw bytes of the RawArr.
+    pub fn as_bytes(&self) -> &[u8] {
+        self.doc.as_bytes()
+    }
+}
+
+impl std::fmt::Debug for RawArr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RawArr")
+            .field("data", &hex::encode(self.doc.as_bytes()))
+            .finish()
+    }
+}
+
+impl TryFrom<&RawArr> for Vec<Bson> {
+    type Error = Error;
+
+    fn try_from(arr: &RawArr) -> Result<Vec<Bson>> {
+        arr.into_iter()
+            .map(|result| {
+                let rawbson = result?;
+                Bson::try_from(rawbson)
+            })
+            .collect()
+    }
+}
+
+impl<'a> IntoIterator for &'a RawArr {
+    type IntoIter = RawArrIter<'a>;
+    type Item = Result<RawBson<'a>>;
+
+    fn into_iter(self) -> RawArrIter<'a> {
+        RawArrIter {
+            inner: self.doc.into_iter(),
+        }
+    }
+}
+
+/// An iterator over borrowed raw BSON array values.
+pub struct RawArrIter<'a> {
+    inner: Iter<'a>,
+}
+
+impl<'a> Iterator for RawArrIter<'a> {
```

Comment:
One validation we're missing here is checking to make sure that the keys in the array are well-formed (i.e. incrementing integers starting with 0). This could lead to some confusing behavior in which malformed BSON bytes that include an "\x04" tag followed by a regular document are successfully interpreted as an array. I think we should add some checks here to ensure that the document we're dealing with is a proper array-style document.

Reply:
Surprisingly, accepting array documents with messed up indexes is actually required by the bson corpus tests. e.g. the "Single Element array with index set incorrectly to ab" test is included under "valid": https://github.com/mongodb/specifications/blob/master/source/bson-corpus/tests/array.json#L23

Reply:
Huh, that's interesting.